src/psyclone/domain/lfric/lfric_stencils.py

```diff
@@ -0,0 +1,649 @@
+        # to this invoke routine from the algorithm layer. Duplicate
+        # argument names are removed.
+        self._unique_direction_args = []
+        direction_names = []
+        for call in self._calls:
+            for idx, arg in enumerate(call.arguments.args):
+                if arg.stencil and arg.stencil.direction_arg:
+                    if arg.stencil.direction_arg.is_literal():
+                        raise GenerationError(
+                            f"Kernel {call.name}, metadata arg {idx}, a "
+                            f"literal is not a valid value for a stencil "
+                            f"direction")
+                    if arg.stencil.direction_arg.text.lower() not in \
+                       ["x_direction", "y_direction"]:
+                        if arg.stencil.direction_arg.text not in \
+                           direction_names:
+                            direction_names.append(
+                                arg.stencil.direction_arg.text)
+                            self._unique_direction_args.append(arg)
+
+        # list of stencil args with an extent variable passed in. The same
+        # field name may occur more than once here from different kernels.
+        self._kern_args = []
+        for call in self._calls:
+            for arg in call.arguments.args:
+                if arg.stencil:
+                    if not arg.stencil.extent:
+                        self._kern_args.append(arg)
+
+    @staticmethod
+    def extent_value(arg):
+        '''
+        Returns the content of the stencil extent which may be a literal
+        value (a number) or a variable name. This function simplifies this
+        problem by returning a string in either case.
+
+        :param arg: the argument with which the stencil is associated.
+        :type arg: :py:class:`psyclone.dynamo0p3.DynKernelArgument`
+
+        :returns: the content of the stencil extent.
+        :rtype: str
+
+        '''
+        if arg.stencil.extent_arg.is_literal():
+            return arg.stencil.extent_arg.text
+        return arg.stencil.extent_arg.varname
+
+    @staticmethod
+    def stencil_unique_str(arg, context):
+        '''
+        Creates a unique identifier for a stencil. As a stencil
+        differs due to the function space it operates on, type of
+        stencil and extent of stencil, we concatenate these things together
+        to create a unique string.
+
+        :param arg: kernel argument with which stencil is associated.
+        :type arg: :py:class:`psyclone.dynamo0p3.DynKernelArgument`
+        :param str context: a context for this stencil (e.g. "size" or \
+                            "direction").
+
+        :returns: unique string identifying the stencil for this argument.
+        :rtype: str
+
+        :raises GenerationError: if an explicit stencil extent is found in \
+                                 the meta-data for the kernel argument.
```

Comment:
"metadata" here and throughout the file, I think. Please also remove `\` from descriptions and add a blank line above the triple quotes as said above.